src/blr/cli/tagbam.py

```diff
@@ -1,58 +1,82 @@
 """
-Transfers barcode sequence information from the input file alignment name to SAM tags in output file.
+Strips headers from tags and depending on mode, set the appropriate SAM tag.
 """
 
 import pysam
 import logging
 from tqdm import tqdm
+from collections import Counter
+
+from blr.utils import print_stats
 
 logger = logging.getLogger(__name__)
 
 
 def main(args):
+    # Can't be at top since function are defined later
+    FUNCTION_DICT = {
+        "sam": mode_samtags_underline_separation,
+        "ema": mode_ema
+    }
 
-    # Generate dict with bc => bc_cluster consensus sequence
     logger.info("Starting analysis")
-    alignments_missing_bc = 0
+    summary = Counter()
+    processing_function = FUNCTION_DICT[args.format]
 
     # Read SAM/BAM files and transfer barcode information from alignment name to SAM tag
     with pysam.AlignmentFile(args.input, "rb") as infile, \
             pysam.AlignmentFile(args.output, "wb", template=infile) as out:
+        for read in tqdm(infile.fetch(until_eof=True), desc="Processing reads", unit=" reads"):
+            # Strips header from tag and depending on script mode, possibly sets SAM tag
+            processing_function(read, summary)
+            out.write(read)
 
-        for read in tqdm(infile.fetch(until_eof=True), desc="Reading input"):
-            try:
-                name, raw_barcode_tag, corr_barcode_tag = read.query_name.split()[0].split('_')
-            except ValueError:
-                alignments_missing_bc += 1
-                name, raw_barcode_tag, corr_barcode_tag = read.query_name.split()[0], None, None
+    print_stats(summary, name=__name__)
+    logger.info("Finished")
 
-            # Rename aligment to exclude barcode information.
-            read.query_name = name
 
-            if corr_barcode_tag:
-                assert corr_barcode_tag.startswith(f"{args.barcode_tag}:Z:")
-                corr_barcode = corr_barcode_tag.split(":")[-1]
-                read.set_tag(args.barcode_tag, corr_barcode, value_type='Z')
+def mode_samtags_underline_separation(read, summary):
+    """
+    Trims header from tags and sets SAM tags according to values found in header.
+    Assumes format: @header_<tag>:<type>:<seq> (can be numerous tags). Constrictions are: Header includes SAM tags
+    separated by "_".
+    :param read: pysam read alignment
+    :param summary: Collections's Counter object
+    :return:
+    """
 
-            if raw_barcode_tag:
-                assert raw_barcode_tag.startswith(f"{args.sequence_tag}:Z:")
-                raw_barcode = raw_barcode_tag.split(":")[-1]
-                read.set_tag(args.sequence_tag, raw_barcode, value_type='Z')
+    # Strip header
+    summary["Total reads"] += 1
```

Comment:
I would move this statistic to the main function as it is the same independent of the function used.